language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -47,63 +47,66 @@ optionsParserInfo = info (optionsParser <**> helper)
     <> progDesc "Generate TypeScript bindings from a DAR"
     )
 
+-- Each package has a unique ID and is perhaps ascribed a symbolic
+-- name (that is, if a given package has been determined to be the
+-- "main-package" of some DAR).
+type PackageMap = Map.Map PackageId (Maybe String, Package)
+
+-- Merge a DAR into a package map (two DARs may have one or more
+-- packages in common). Merging fails and an IO exception is raised
+-- if:
+--   * Two packages are ascribed the same symbolic name;
+--   * A single package is ascribed two different names.
+mergePackageMap :: PackageMap -> FilePath -> IO PackageMap
+mergePackageMap pkgs dar = do
+  dar <- B.readFile dar
+  let archive = Zip.toArchive $ BSL.fromStrict dar
+  dalfs <- either fail pure $ DAR.readDalfs archive
+  DAR.DalfManifest{packageName} <- either fail pure $ DAR.readDalfManifest archive
+  let allDalfsInDar = (DAR.mainDalf dalfs, packageName) : map (, Nothing) (DAR.dalfs dalfs)
+  foldM merge pkgs allDalfsInDar
+  where
+    merge :: PackageMap -> (BSL.ByteString, Maybe String) -> IO PackageMap
+    merge pkgs (dalf, mbPkgName) = do
+      (pkgId, pkg) <- either (fail . show)  pure $ Archive.decodeArchive Archive.DecodeAsMain (BSL.toStrict dalf)
+      -- Check for errors.
+      let pkgNames = mapMaybe fst (Map.elems pkgs)
+      maybe (maybe (pure ())
```

Comment:
I find this part very hard to read. `maybe (pure ()) f mb ` should be equivalent to `whenJust mb f` but this might just be easier to read with a pattern match.

Reply:
Ok